pkg/provisioner/ironic/updatehardwarestate_test.go

```diff
@@ -0,0 +1,141 @@
+
+	"github.com/gophercloud/gophercloud/openstack/baremetal/v1/nodes"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestUpdateHardwareState(t *testing.T) {
+
+	nodeUUID := "33ce8659-7400-4c68-9535-d10766f07a58"
+
+	cases := []struct {
+		name                 string
+		ironic               *testserver.IronicMock
+		inspector            *testserver.InspectorMock
+		hostCurrentlyPowered bool
+
+		expectedDirty        bool
+		expectedRequestAfter int
+		expectedResultError  string
+
+		expectedPublish string
+		expectedError   string
+	}{
+		{
+			name: "unkown-power-state",
+			ironic: testserver.NewIronic(t).Ready().WithNode(nodes.Node{
+				UUID: nodeUUID,
+			}),
+		},
+		{
+			name: "updated-power-on-state",
+			ironic: testserver.NewIronic(t).Ready().WithNode(nodes.Node{
+				UUID:       nodeUUID,
+				PowerState: "power on",
+			}),
+			hostCurrentlyPowered: true,
+		},
+		{
+			name: "not-updated-power-on-state",
+			ironic: testserver.NewIronic(t).Ready().WithNode(nodes.Node{
+				UUID:       nodeUUID,
+				PowerState: "power on",
+			}),
+			hostCurrentlyPowered: false,
+
+			expectedDirty: true,
+		},
+		{
+			name: "updated-power-off-state",
+			ironic: testserver.NewIronic(t).Ready().WithNode(nodes.Node{
+				UUID:       nodeUUID,
+				PowerState: "power off",
+			}),
+			hostCurrentlyPowered: false,
+		},
+		{
+			name: "not-updated-power-off-state",
+			ironic: testserver.NewIronic(t).Ready().WithNode(nodes.Node{
+				UUID:       nodeUUID,
+				PowerState: "power off",
+			}),
+			hostCurrentlyPowered: true,
+
+			expectedDirty: true,
+		},
+		{
+			name: "no-power",
+			ironic: testserver.NewIronic(t).Ready().WithNode(nodes.Node{
+				UUID:       nodeUUID,
+				PowerState: "None",
+			}),
+			hostCurrentlyPowered: true,
+		},
+		{
+			name:   "not-existing-host",
+			ironic: testserver.NewIronic(t).Ready(),
+
+			expectedError: "failed to find existing host: failed to find node by ID 33ce8659-7400-4c68-9535-d10766f07a58: EOF",
+		},
+		{
+			name:   "node-not-found",
+			ironic: testserver.NewIronic(t).Ready().WithNodeError(nodeUUID, http.StatusNotFound),
+
+			expectedError: "failed to find existing host: failed to find node by name myhost: EOF",
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			if tc.ironic != nil {
+				tc.ironic.Start()
+				defer tc.ironic.Stop()
+			}
+
+			if tc.inspector != nil {
+				tc.inspector.Start()
+				defer tc.inspector.Stop()
+			}
+
+			host := makeHost()
+			host.Status.PoweredOn = tc.hostCurrentlyPowered
+
+			publishedMsg := ""
+			publisher := func(reason, message string) {
+				publishedMsg = reason + " " + message
+			}
+			auth := clients.AuthConfig{Type: clients.NoAuth}
+			prov, err := newProvisionerWithSettings(host, bmc.Credentials{}, publisher,
+				tc.ironic.Endpoint(), auth, tc.inspector.Endpoint(), auth,
+			)
+			if err != nil {
+				t.Fatalf("could not create provisioner: %s", err)
+			}
+
+			prov.status.ID = nodeUUID
```

Comment:
This should be set on host after it is created on line 109.

Reply:
Thanks, good catch